File: src/converter/plugins/convertConstructorDeclaration.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {convertParameterDeclarations} from "./convertParameterDeclaration.js";

export const convertConstructorDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isConstructorDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    return convertParameterDeclarations(node, context, render, {
        strategy: "function",
        template: parameters => {
            return `constructor (${parameters})`
        }
    })
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertConstructorDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isConstructorDeclaration

val convertConstructorDeclaration = createPlugin plugin@{ node, context, render ->
    if (!isConstructorDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    convertParameterDeclarations(node, context, render, ParameterDeclarationsConfiguration(
        strategy = ParameterDeclarationStrategy.function,
        template = { parameters, _ ->
            "constructor (${parameters})"
        }
    ))
}
